client/components/d3-base/index.js

```diff
@@ -0,0 +1,131 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import classNames from 'classnames';
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import { isEmpty } from 'lodash';
+import { select as d3Select } from 'd3-selection';
+
+/**
+ * Provides foundation to use D3 within React.
+ *
+ * React is responsible for determining when a chart should be updated (e.g. whenever data changes or the browser is
+ * resized), while D3 is responsible for the actual rendering of the chart (which is performed via DOM operations that
+ * happen outside of React's control).
+ *
+ * This component makes use of new lifecycle methods that come with React 16.3. Thus, while this component (i.e. the
+ * container of the chart) is rendered during the 'render phase' the chart itself is only rendered during the 'commit
+ * phase' (i.e. in 'componentDidMount' and 'componentDidUpdate' methods).
+ */
+export default class D3Base extends Component {
+	static propTypes = {
+		className: PropTypes.string,
+		data: PropTypes.any, // required to detect changes in data
+		drawChart: PropTypes.func.isRequired,
+		getParams: PropTypes.func.isRequired,
+	};
+
+	state = {
+		data: null,
+		params: null,
+		drawChart: null,
+		getParams: null,
+	};
+
+	chartRef = React.createRef();
+
+	/**
+	 * Determines if new props were provided, and updates the local state accordingly. This basically stores any new
+	 * prop in the local state (so it can be used for comparison in future calls), and resets the list of params in that
+	 * case (in order to indicate that the chart should be drawn again).
+	 */
+	static getDerivedStateFromProps( nextProps, prevState ) {
+		let state = {};
+
+		if ( nextProps.data !== prevState.data ) {
+			state = { ...state, data: nextProps.data };
+		}
+
+		if ( nextProps.drawChart !== prevState.drawChart ) {
```

Comment:
`drawChart` and `getParams` should not be allowed to change I think. We should spend a bit more time on D3Base when we come back to it I think